File: api/App.Api/Models/DocumentModels.cs
namespace App.Api.Models;

public class DocumentResponse
{
    public int Id { get; set; }
    public int? ContactId { get; set; }
    public string FileName { get; set; } = default!;
    public string ContentType { get; set; } = default!;
    public long SizeBytes { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

public class CreateUploadUrlRequest
{
    public int? ContactId { get; set; }
    public string FileName { get; set; } = default!;
    public string ContentType { get; set; } = default!;
}

public class CreateUploadUrlResponse
{
    public string UploadUrl { get; set; } = default!;
    public string StorageKey { get; set; } = default!;
}

public class CompleteUploadRequest
{
    public int? ContactId { get; set; }
    public string FileName { get; set; } = default!;
    public string ContentType { get; set; } = default!;
    public long SizeBytes { get; set; }
    public string StorageKey { get; set; } = default!;
}

public class DownloadUrlResponse
{
    public string DownloadUrl { get; set; } = default!;
}

File: web/app/lib/document-models.ts
export type Document = {
  id: number;
  contactId: number | null;
  fileName: string;
  contentType: string;
  sizeBytes: number;
  createdAt: string;
  updatedAt: string;
};

export type CreateUploadUrlResponse = {
  uploadUrl: string;
  storageKey: string;
  uploadHeaders: Record<string, string>;
};

export type DownloadUrlResponse = {
  downloadUrl: string;
};

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) {
    return `${bytes} B`;
  }

  const units = ["KB", "MB", "GB", "TB"];
  let value = bytes / 1024;
  let unitIndex = 0;

  while (value >= 1024 && unitIndex < units.length - 1) {
    value /= 1024;
    unitIndex += 1;
  }

  return `${value.toFixed(1)} ${units[unitIndex]}`;
}
